### DISCUSSION TITLE
Column Name as a Variable

### DISCUSSION THREAD

**Discussion by JerContact** [simolus3/drift#2373] [Q&A]
I tried looking but I seem to be unable to find an answer on this one, if I'm trying to do this in the drift file:

searchMainAscending: SELECT Filename, FilePath
 FROM MAIN WHERE MAIN.PK IN ( 
    SELECT FTS5.rowid FROM FTS5 WHERE FTS5 MATCH (?) LIMIT (?) OFFSET (?)
 ) ORDER BY (?) COLLATE NOCASE;

For the "ORDER BY (?)" I want to pass in a variable that is the column header (so either Filename or FilePath), but no matter what i do I can't seem to get this statement to work in drift.  Even if I do lowercase string.  Is there a way to have this work like this or do i need to incorperate something into the .dart file?

**Comment by simolus3** [comment]
Generally,  SQL variables can't do this. Variables are always bound to direct values, so they can only replace literals. Passing a lowercase string will just do `ORDER BY ('filename')` (note the quotes) which does not actually read the column from the table.

As an alternative to variables, drift files also support [Dart components](https://drift.simonbinder.eu/docs/using-sql/drift_files/#dart-interop) in queries. This is a drift-specific SQL extension that allows you to embed SQL snippets created with drift's Dart API in queries defined in drift file. For instance, if you do

```sql
searchMainAscending: SELECT Filename, FilePath
FROM MAIN WHERE MAIN.PK IN (
SELECT FTS5.rowid FROM FTS5 WHERE FTS5 MATCH (?) LIMIT (?) OFFSET (?)
) ORDER BY $expr COLLATE NOCASE;
```

Then drift will generate a method which you can call with `searchMainAscending(expr: (main) => main.filename)`. That can be used to select the column (or any other expression to use in the ordering term) as needed.
